module/applications/calendar/calendar-hud.mjs

```diff
@@ -0,0 +1,280 @@
+  async _preparePartContext(partId, context, options) {
+    context = await super._preparePartContext(partId, context, options);
+    switch ( partId ) {
+      case "endButtons":
+        context.buttons = context.buttons.filter(b => b.position === "end");
+        break;
+      case "startButtons":
+        context.buttons = context.buttons.filter(b => b.position === "start").reverse();
+        break;
+    }
+    return context;
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Adjust the date, time, and progress in the core part without a full re-render to allow animation.
+   * @param {CalendarTimeDeltas} [deltas={}]  Information on the time change deltas.
+   */
+  async renderCore(deltas={}) {
+    const prefs = game.settings.get("dnd5e", "calendarPreferences");
+    const dateFormatter = CONFIG.DND5E.calendar.formatters.find(f => f.value === prefs.formatters.date);
+    this.element.querySelector(".calendar-date").innerText = dateFormatter ? game.time.calendar.format(
+      game.time.components, dateFormatter.formatter
+    ) : "";
+    const timeFormatter = CONFIG.DND5E.calendar.formatters.find(f => f.value === prefs.formatters.time);
+    this.element.querySelector(".calendar-time").innerText = timeFormatter ? game.time.calendar.format(
+      game.time.components, timeFormatter.formatter
+    ) : "";
+
+    const widget = this.element.querySelector(".calendar-widget");
+    const adjustProgress = (variable, type, delta=0) => {
+      const currentProgress = Number(widget.style.getPropertyValue(variable));
+      const modulo = (n, d) => ((n % d) + d) % d;
+      const partialProgress = modulo((currentProgress + 0.5), 2) - 0.5;
+      const targetProgress = type in game.time.calendar ? game.time.calendar[type]()
+        : CalendarData5e.prototype[type].call(game.time.calendar);
+      const difference = targetProgress - partialProgress;
+      const newProgress = currentProgress + difference + (delta * 2);
+      widget.style.setProperty(variable, newProgress);
+    };
+    adjustProgress("--calendar-day-progress", "progressDay", deltas.midnights);
+    adjustProgress("--calendar-night-progress", "progressNight", deltas.middays);
+  }
+
+  /* -------------------------------------------- */
+  /*  Life-Cycle Handlers                         */
+  /* -------------------------------------------- */
+
+  /** @inheritDoc */
+  async _onRender(context, options) {
+    await super._onRender(context, options);
+    await this.renderCore();
+
+    const addButtonEvent = (button, config) => {
+      if ( typeof config?.onClick === "function" ) {
+        button.addEventListener("click", event => {
+          event.preventDefault();
+          config.onClick(event);
+        });
+      }
+    };
+
+    for ( const button of this.element.querySelectorAll(".calendar-buttons > .calendar-button > button") ) {
+      const config = context.buttons[parseInt(button.dataset.index)];
+      addButtonEvent(button, config);
```

Comment:
I suppose you have to do this in `_onRender` because you need `context.buttons`. The ways we handled something like this elsewhere (like journal sheets, for example), is to cache the prepared context so that it can be accessed outside of the normal render pipeline.

For example, in `_prepareButtonsContext`, you could have something like:
```js
this.#buttons = context.buttons = /* ... */
```
That would allow you to handle these button clicks in `_onClickAction` without having to bind your own listeners.